File: nanopay/src/net/nanopay/meter/LogoutDisabledUserDAO.java
package net.nanopay.meter;

import foam.core.FObject;
import foam.core.X;
import foam.dao.ArraySink;
import foam.dao.DAO;
import foam.dao.ProxyDAO;
import foam.mlang.MLang;
import foam.nanos.auth.AuthService;
import foam.nanos.auth.User;
import foam.nanos.session.Session;
import foam.util.SafetyUtil;
import net.nanopay.model.Business;

import java.util.List;

public class LogoutDisabledUserDAO extends ProxyDAO {
  protected DAO sessionDAO_;
  protected AuthService auth_;

  public LogoutDisabledUserDAO(X x, DAO delegate) {
    super(x, delegate);
  }

  @Override
  public FObject put_(X x, FObject obj) {
    User newUser = (User) obj;
    User oldUser = (User) getDelegate().find(newUser.getId());

    if (
      oldUser != null
      && oldUser.getEnabled()
      && !newUser.getEnabled()
    ) {
      sessionDAO_ = (DAO) x.get("sessionDAO");
      auth_ = (AuthService) x.get("auth");

      if (SafetyUtil.equals(newUser.getGroup(), "sme")) {
        logoutSmeUser(newUser, newUser.getEntities(x).getDAO());
      } else {
        logout(newUser, null);
      }
    }

    return super.put_(x, obj);
  }

  private void logoutSmeUser(User user, DAO entitiesDao) {
    ArraySink sink = (ArraySink) entitiesDao.select(new ArraySink());
    List<Business> businesses = sink.getArray();
    for (Business business : businesses) {
      logout(user, business);
    }
  }

  private void logout(User user, Business business) {
    long userId = user.getId();
    ArraySink sink = (ArraySink) sessionDAO_.where(
      MLang.EQ(Session.USER_ID,
        business != null ? business.getId() : userId))
      .select(new ArraySink());
    List<Session> sessions = sink.getArray();

    for (Session session : sessions) {
      User agent = (User) session.getContext().get("agent");
      if (
        session.getUserId() == userId
        || (agent !=null && agent.getId() == userId)
      ) {
        auth_.logout(session.getContext());
      }
    }
  }
}

File: nanopay/src/net/nanopay/security/auth/LogoutDisabledUserDAO.js
/**
 * NANOPAY CONFIDENTIAL
 *
 * [2020] nanopay Corporation
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of nanopay Corporation.
 * The intellectual and technical concepts contained
 * herein are proprietary to nanopay Corporation
 * and may be covered by Canadian and Foreign Patents, patents
 * in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from nanopay Corporation.
 */

foam.CLASS({
  package: 'net.nanopay.security.auth',
  name: 'LogoutDisabledUserDAO',
  extends: 'foam.dao.ProxyDAO',

  documentation: `DAO decorator that logs out the user who is being disabled.`,

  javaImports: [
    'foam.core.Detachable',
    'foam.dao.AbstractSink',
    'foam.dao.DAO',
    'foam.mlang.MLang',
    'foam.nanos.auth.AuthService',
    'foam.nanos.auth.Subject',
    'foam.nanos.auth.User',
    'foam.nanos.session.Session',

    'net.nanopay.admin.model.AccountStatus'
  ],

  methods: [
    {
      name: 'put_',
      javaCode: `
        User newUser = (User) obj;
        User oldUser = (User) getDelegate().find(newUser.getId());

        if (
          oldUser != null
          && AccountStatus.DISABLED != oldUser.getStatus()
          && AccountStatus.DISABLED == newUser.getStatus()
        ) {
          sessionDAO_ = (DAO) x.get("localSessionDAO");
          auth_ = (AuthService) x.get("auth");

          logoutUser(newUser, null);
          logoutAgent(newUser, newUser.getEntities(x).getDAO());
        }

        return super.put_(x, obj);
      `
    },
    {
      name: 'logoutAgent',
      args: [
        { type: 'foam.nanos.auth.User', name: 'agent' },
        { type: 'foam.dao.DAO', name: 'entitiesDAO'}
      ],
      javaCode: `
        entitiesDAO.select(new AbstractSink() {
          @Override
          public void put(Object obj, Detachable sub) {
            logoutUser(agent, (User) obj);
          }
        });
      `
    },
    {
      name: 'logoutUser',
      args: [
        { type: 'foam.nanos.auth.User', name: 'user' },
        { type: 'foam.nanos.auth.User', name: 'entity' }
      ],
      javaCode: `
        long userId = user.getId();
        sessionDAO_.where(
          MLang.EQ(Session.USER_ID, entity != null ? entity.getId() : userId)
        ).select(new AbstractSink() {
          @Override
          public void put(Object obj, Detachable sub) {
            Session session = (Session) obj;
            User agent = ((Subject) session.getContext().get("subject")).getRealUser();
            if ( session.getUserId() == userId
              || (agent != null && agent.getId() == userId)
            ) {
              auth_.logout(session.getContext());
            }
          }
        });
      `
    }
  ],

  axioms: [
    {
      name: 'javaExtras',
      buildJavaClass: function(cls) {
        cls.extras.push(`
          private DAO sessionDAO_;
          private AuthService auth_;
        `);
      }
    }
  ],
});
